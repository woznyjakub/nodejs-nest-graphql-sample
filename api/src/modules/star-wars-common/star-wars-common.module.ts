import { Module } from '@nestjs/common';

import { StarWarsCommonService } from './star-wars-common.service';

@Module({
  providers: [StarWarsCommonService],
  exports: [StarWarsCommonService],
})
export class StarWarsCommonModule {}
