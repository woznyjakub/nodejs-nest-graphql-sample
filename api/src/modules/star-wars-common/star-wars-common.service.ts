import { CACHE_MANAGER } from '@nestjs/cache-manager';
import { Inject, Injectable } from '@nestjs/common';
import { Cache } from 'cache-manager';

export type PagedResponse<Item> = {
  count: number;
  next: string | null;
  previous: string | null;
  results: Item[];
};

@Injectable()
export class StarWarsCommonService {
  constructor(@Inject(CACHE_MANAGER) private cacheManager: Cache) {}

  async fetchWithCaching<T extends object>(
    cacheKey: string,
    getterFn: () => Promise<T>,
  ): Promise<T> {
    const cachedData = await this.cacheManager.get<T>(cacheKey);
    if (cachedData) {
      return cachedData;
    }

    const data = await getterFn();
    await this.cacheManager.set(cacheKey, data);

    return data;
  }

  async getAllPages<T extends object>(
    cacheHandlerFn: (page: number) => string,
    getterFn: (page: number) => Promise<PagedResponse<T>>,
  ): Promise<T[]> {
    const limit = 25;
    let page = 1;

    const results: T[] = [];

    do {
      const data = await this.fetchWithCaching(cacheHandlerFn(page), () => getterFn(page));

      if (Array.isArray(data.results) || !results?.length) {
        results.push(...data.results);
      } else {
        break;
      }

      if (!data.next) {
        break;
      }
      page++;
    } while (page < limit);

    return results;
  }

  private buildCacheKey(
    resourceName: string,
    dataType: 'page' | 'id',
    value: string | number,
  ): string {
    return `swapi-${resourceName}-${dataType}-${value}`;
  }

  getVehicleCacheKey(id: number): string {
    return this.buildCacheKey('vehicle', 'id', id);
  }

  getVehiclesCacheKey(page: number): string {
    return this.buildCacheKey('vehicles', 'page', page);
  }

  getFilmCacheKey(id: number): string {
    return this.buildCacheKey('film', 'id', id);
  }

  getFilmsCacheKey(page: number): string {
    return this.buildCacheKey('films', 'page', page);
  }

  getStarshipCacheKey(id: number): string {
    return this.buildCacheKey('starship', 'id', id);
  }

  getStarshipsCacheKey(page: number): string {
    return this.buildCacheKey('starships', 'page', page);
  }

  getPlanetCacheKey(id: number): string {
    return this.buildCacheKey('planet', 'id', id);
  }

  getPlanetsCacheKey(page: number): string {
    return this.buildCacheKey('planets', 'page', page);
  }

  getSpeciesCacheKey(id: number): string {
    return this.buildCacheKey('species', 'id', id);
  }

  getSpeciesManyCacheKey(page: number): string {
    return this.buildCacheKey('speciesMany', 'page', page);
  }

  getPeopleCacheKey(page: number): string {
    return this.buildCacheKey('people', 'page', page);
  }
}
