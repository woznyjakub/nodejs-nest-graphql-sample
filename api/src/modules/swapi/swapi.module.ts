import { HttpModule } from '@nestjs/axios';
import { Module } from '@nestjs/common';

import { SwapiService } from './swapi.service';

@Module({
  imports: [HttpModule],
  providers: [SwapiService],
  exports: [SwapiService],
})
export class SwapiModule {}
