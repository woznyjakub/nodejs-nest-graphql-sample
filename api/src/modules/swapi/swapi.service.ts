import { HttpService } from '@nestjs/axios';
import { HttpException, Injectable } from '@nestjs/common';
import { ClassConstructor, plainToInstance } from 'class-transformer';
import { validate } from 'class-validator';
import { lastValueFrom } from 'rxjs';

import { FilmResponse, FilmsResponse } from './dto/film';
import { PeopleResponse } from './dto/people';
import { PlanetResponse, PlanetsResponse } from './dto/planet';
import { SpeciesManyResponse, SpeciesResponse } from './dto/species';
import { StarshipResponse, StarshipsResponse } from './dto/starship';
import { VehicleResponse, VehiclesResponse } from './dto/vehicle';

@Injectable()
export class SwapiService {
  constructor(private readonly httpService: HttpService) {}

  private readonly baseUrl = 'https://swapi.dev/api';

  private async getResource<T extends object>(
    name: string,
    dto: ClassConstructor<T>,
    id?: number,
    page?: number,
  ): Promise<T> {
    const url = `${this.baseUrl}/${name}${id ? `/${id}` : ''}`;
    const params = !id && page ? { page } : null;

    try {
      const response = await lastValueFrom(this.httpService.get(url, { params }));
      const resourceData: unknown = response.data;

      const resource = plainToInstance<T, unknown>(dto, resourceData, {
        enableImplicitConversion: true,
        excludeExtraneousValues: true,
      });

      const errors = await validate(resource);
      if (errors.length > 0) {
        throw new Error(
          `Swapi resource validation failed: ${errors.map((e) => e.toString()).join(', ')}`,
        );
      }

      return resource;
    } catch (error) {
      throw new HttpException(
        `Failed to get swapi resource (${name}). Error: ${error.message}`,
        error.status,
      );
    }
  }

  getVehicle(id: number): Promise<VehicleResponse> {
    return this.getResource('vehicles', VehicleResponse, id);
  }

  getVehicles(page: number = 1): Promise<VehiclesResponse> {
    return this.getResource('vehicles', VehiclesResponse, undefined, page);
  }

  getFilm(id: number): Promise<FilmResponse> {
    return this.getResource('films', FilmResponse, id);
  }

  getFilms(page: number = 1): Promise<FilmsResponse> {
    return this.getResource('films', FilmsResponse, undefined, page);
  }

  getStarship(id: number): Promise<StarshipResponse> {
    return this.getResource('starships', StarshipResponse, id);
  }

  getStarships(page: number = 1): Promise<StarshipsResponse> {
    return this.getResource('starships', StarshipsResponse, undefined, page);
  }

  getPlanet(id: number): Promise<PlanetResponse> {
    return this.getResource('planets', PlanetResponse, id);
  }

  getPlanets(page: number = 1): Promise<PlanetsResponse> {
    return this.getResource('planets', PlanetsResponse, undefined, page);
  }

  getSpecies(id: number): Promise<SpeciesResponse> {
    return this.getResource('species', SpeciesResponse, id);
  }

  getSpeciesMany(page: number = 1): Promise<SpeciesManyResponse> {
    return this.getResource('species', SpeciesManyResponse, undefined, page);
  }

  getPeople(page: number = 1): Promise<PeopleResponse> {
    return this.getResource('people', PeopleResponse, undefined, page);
  }
}
