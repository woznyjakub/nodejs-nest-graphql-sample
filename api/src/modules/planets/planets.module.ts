import { Module } from '@nestjs/common';

import { StarWarsCommonModule } from '../star-wars-common/star-wars-common.module';
import { SwapiModule } from '../swapi/swapi.module';

import { PlanetsResolver } from './planets.resolver';
import { PlanetsService } from './planets.service';

@Module({
  imports: [SwapiModule, StarWarsCommonModule],
  providers: [PlanetsResolver, PlanetsService],
})
export class PlanetsModule {}
