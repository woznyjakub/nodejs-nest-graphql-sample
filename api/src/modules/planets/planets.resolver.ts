import { Resolver, Query, Args, Int } from '@nestjs/graphql';

import { Planet } from './entities/planet.entity';
import { Planets } from './entities/planets.entity';
import { PlanetsService } from './planets.service';

@Resolver(() => Planet)
export class PlanetsResolver {
  constructor(private readonly vehiclesService: PlanetsService) {}

  @Query(() => Planets, { name: 'planets' })
  findAll(@Args('page', { type: () => Int }) page: number): Promise<Planets> {
    return this.vehiclesService.findAll(page);
  }

  @Query(() => Planet, { name: 'planet' })
  findOne(@Args('id', { type: () => Int }) id: number): Promise<Planet> {
    return this.vehiclesService.findOne(id);
  }
}
