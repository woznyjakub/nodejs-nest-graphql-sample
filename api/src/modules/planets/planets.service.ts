import { Injectable } from '@nestjs/common';

import { StarWarsCommonService } from '../star-wars-common/star-wars-common.service';
import { PlanetResponse } from '../swapi/dto/planet';
import { SwapiService } from '../swapi/swapi.service';

import { Planet } from './entities/planet.entity';
import { Planets } from './entities/planets.entity';

@Injectable()
export class PlanetsService {
  constructor(
    private readonly swapiService: SwapiService,
    private readonly swCommonService: StarWarsCommonService,
  ) {}

  async findAll(page: number): Promise<Planets> {
    const cacheKey = `swapi-planets-page-${page}`;

    const swapiData = await this.swCommonService.fetchWithCaching(cacheKey, () =>
      this.swapiService.getPlanets(page),
    );

    return {
      count: swapiData.count,
      next: swapiData.next,
      previous: swapiData.previous,
      results: swapiData.results.map(this.mapSinglePlanet),
    };
  }

  async findOne(id: number): Promise<Planet> {
    const cacheKey = `swapi-planets-id-${id}`;

    const swapiData = await this.swCommonService.fetchWithCaching(cacheKey, () =>
      this.swapiService.getPlanet(id),
    );

    return this.mapSinglePlanet(swapiData);
  }

  private mapSinglePlanet(planet: PlanetResponse): Planet {
    return {
      name: planet.name,
      rotationPeriod: planet.rotation_period,
      orbitalPeriod: planet.orbital_period,
      diameter: planet.diameter,
      climate: planet.climate,
      gravity: planet.gravity,
      terrain: planet.terrain,
      surfaceWater: planet.surface_water,
      population: planet.population,
      residents: planet.residents,
      films: planet.films,
      url: planet.url,
    };
  }
}
