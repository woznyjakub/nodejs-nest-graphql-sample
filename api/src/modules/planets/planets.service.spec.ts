import { Test, TestingModule } from '@nestjs/testing';

import { planetResponseMock, planetsResponseMock } from '../../test/mocks/swapi';
import { StarWarsCommonService } from '../star-wars-common/star-wars-common.service';
import { SwapiService } from '../swapi/swapi.service';

import { PlanetsService } from './planets.service';

describe('PlanetsService', () => {
  let planetsService: PlanetsService;

  const mockFetchWithCaching = vi.fn();

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        {
          provide: StarWarsCommonService,
          useValue: {
            fetchWithCaching: mockFetchWithCaching,
          },
        },
        PlanetsService,
        {
          provide: SwapiService,
          useValue: {
            getPlanet: vi.fn(),
            getPlanets: vi.fn(),
          },
        },
      ],
    }).compile();

    planetsService = module.get<PlanetsService>(PlanetsService);
  });

  describe('single planet data', () => {
    it('should return mapped data', async () => {
      mockFetchWithCaching.mockReturnValue(planetResponseMock);

      const result = await planetsService.findOne(1);

      expect(result).toMatchSnapshot();
    });
  });

  describe('multiple planets data', () => {
    it('should return mapped data when', async () => {
      mockFetchWithCaching.mockReturnValue(planetsResponseMock);

      const result = await planetsService.findAll(1);

      expect(result).toMatchSnapshot();
    });
  });
});
