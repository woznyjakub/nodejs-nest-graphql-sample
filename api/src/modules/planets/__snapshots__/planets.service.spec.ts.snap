// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PlanetsService > multiple planets data > should return mapped data when 1`] = `
{
  "count": 60,
  "next": "https://swapi.dev/api/planets/?page=2",
  "previous": null,
  "results": [
    {
      "climate": "arid",
      "diameter": "10465",
      "films": [
        "https://swapi.dev/api/films/1/",
        "https://swapi.dev/api/films/3/",
        "https://swapi.dev/api/films/4/",
        "https://swapi.dev/api/films/5/",
        "https://swapi.dev/api/films/6/",
      ],
      "gravity": "1 standard",
      "name": "Tatooine",
      "orbitalPeriod": "304",
      "population": "200000",
      "residents": [
        "https://swapi.dev/api/people/1/",
        "https://swapi.dev/api/people/2/",
        "https://swapi.dev/api/people/4/",
        "https://swapi.dev/api/people/6/",
        "https://swapi.dev/api/people/7/",
        "https://swapi.dev/api/people/8/",
        "https://swapi.dev/api/people/9/",
        "https://swapi.dev/api/people/11/",
        "https://swapi.dev/api/people/43/",
        "https://swapi.dev/api/people/62/",
      ],
      "rotationPeriod": "23",
      "surfaceWater": "1",
      "terrain": "desert",
      "url": "https://swapi.dev/api/planets/1/",
    },
    {
      "climate": "temperate",
      "diameter": "12500",
      "films": [
        "https://swapi.dev/api/films/1/",
        "https://swapi.dev/api/films/6/",
      ],
      "gravity": "1 standard",
      "name": "Alderaan",
      "orbitalPeriod": "364",
      "population": "2000000000",
      "residents": [
        "https://swapi.dev/api/people/5/",
        "https://swapi.dev/api/people/68/",
        "https://swapi.dev/api/people/81/",
      ],
      "rotationPeriod": "24",
      "surfaceWater": "40",
      "terrain": "grasslands, mountains",
      "url": "https://swapi.dev/api/planets/2/",
    },
    {
      "climate": "temperate, tropical",
      "diameter": "10200",
      "films": [
        "https://swapi.dev/api/films/1/",
      ],
      "gravity": "1 standard",
      "name": "Yavin IV",
      "orbitalPeriod": "4818",
      "population": "1000",
      "residents": [],
      "rotationPeriod": "24",
      "surfaceWater": "8",
      "terrain": "jungle, rainforests",
      "url": "https://swapi.dev/api/planets/3/",
    },
    {
      "climate": "frozen",
      "diameter": "7200",
      "films": [
        "https://swapi.dev/api/films/2/",
      ],
      "gravity": "1.1 standard",
      "name": "Hoth",
      "orbitalPeriod": "549",
      "population": "unknown",
      "residents": [],
      "rotationPeriod": "23",
      "surfaceWater": "100",
      "terrain": "tundra, ice caves, mountain ranges",
      "url": "https://swapi.dev/api/planets/4/",
    },
    {
      "climate": "murky",
      "diameter": "8900",
      "films": [
        "https://swapi.dev/api/films/2/",
        "https://swapi.dev/api/films/3/",
        "https://swapi.dev/api/films/6/",
      ],
      "gravity": "N/A",
      "name": "Dagobah",
      "orbitalPeriod": "341",
      "population": "unknown",
      "residents": [],
      "rotationPeriod": "23",
      "surfaceWater": "8",
      "terrain": "swamp, jungles",
      "url": "https://swapi.dev/api/planets/5/",
    },
    {
      "climate": "temperate",
      "diameter": "118000",
      "films": [
        "https://swapi.dev/api/films/2/",
      ],
      "gravity": "1.5 (surface), 1 standard (Cloud City)",
      "name": "Bespin",
      "orbitalPeriod": "5110",
      "population": "6000000",
      "residents": [
        "https://swapi.dev/api/people/26/",
      ],
      "rotationPeriod": "12",
      "surfaceWater": "0",
      "terrain": "gas giant",
      "url": "https://swapi.dev/api/planets/6/",
    },
    {
      "climate": "temperate",
      "diameter": "4900",
      "films": [
        "https://swapi.dev/api/films/3/",
      ],
      "gravity": "0.85 standard",
      "name": "Endor",
      "orbitalPeriod": "402",
      "population": "30000000",
      "residents": [
        "https://swapi.dev/api/people/30/",
      ],
      "rotationPeriod": "18",
      "surfaceWater": "8",
      "terrain": "forests, mountains, lakes",
      "url": "https://swapi.dev/api/planets/7/",
    },
    {
      "climate": "temperate",
      "diameter": "12120",
      "films": [
        "https://swapi.dev/api/films/3/",
        "https://swapi.dev/api/films/4/",
        "https://swapi.dev/api/films/5/",
        "https://swapi.dev/api/films/6/",
      ],
      "gravity": "1 standard",
      "name": "Naboo",
      "orbitalPeriod": "312",
      "population": "4500000000",
      "residents": [
        "https://swapi.dev/api/people/3/",
        "https://swapi.dev/api/people/21/",
        "https://swapi.dev/api/people/35/",
        "https://swapi.dev/api/people/36/",
        "https://swapi.dev/api/people/37/",
        "https://swapi.dev/api/people/38/",
        "https://swapi.dev/api/people/39/",
        "https://swapi.dev/api/people/42/",
        "https://swapi.dev/api/people/60/",
        "https://swapi.dev/api/people/61/",
        "https://swapi.dev/api/people/66/",
      ],
      "rotationPeriod": "26",
      "surfaceWater": "12",
      "terrain": "grassy hills, swamps, forests, mountains",
      "url": "https://swapi.dev/api/planets/8/",
    },
    {
      "climate": "temperate",
      "diameter": "12240",
      "films": [
        "https://swapi.dev/api/films/3/",
        "https://swapi.dev/api/films/4/",
        "https://swapi.dev/api/films/5/",
        "https://swapi.dev/api/films/6/",
      ],
      "gravity": "1 standard",
      "name": "Coruscant",
      "orbitalPeriod": "368",
      "population": "1000000000000",
      "residents": [
        "https://swapi.dev/api/people/34/",
        "https://swapi.dev/api/people/55/",
        "https://swapi.dev/api/people/74/",
      ],
      "rotationPeriod": "24",
      "surfaceWater": "unknown",
      "terrain": "cityscape, mountains",
      "url": "https://swapi.dev/api/planets/9/",
    },
    {
      "climate": "temperate",
      "diameter": "19720",
      "films": [
        "https://swapi.dev/api/films/5/",
      ],
      "gravity": "1 standard",
      "name": "Kamino",
      "orbitalPeriod": "463",
      "population": "1000000000",
      "residents": [
        "https://swapi.dev/api/people/22/",
        "https://swapi.dev/api/people/72/",
        "https://swapi.dev/api/people/73/",
      ],
      "rotationPeriod": "27",
      "surfaceWater": "100",
      "terrain": "ocean",
      "url": "https://swapi.dev/api/planets/10/",
    },
  ],
}
`;

exports[`PlanetsService > single planet data > should return mapped data 1`] = `
{
  "climate": "temperate",
  "diameter": "12500",
  "films": [
    "https://swapi.dev/api/films/1/",
    "https://swapi.dev/api/films/6/",
  ],
  "gravity": "1 standard",
  "name": "Alderaan",
  "orbitalPeriod": "364",
  "population": "2000000000",
  "residents": [
    "https://swapi.dev/api/people/5/",
    "https://swapi.dev/api/people/68/",
    "https://swapi.dev/api/people/81/",
  ],
  "rotationPeriod": "24",
  "surfaceWater": "40",
  "terrain": "grasslands, mountains",
  "url": "https://swapi.dev/api/planets/2/",
}
`;
